control-plane/typeSpecConverter.cpp

```diff
@@ -103,6 +103,16 @@ bool TypeSpecConverter::preorder(const IR::Type_Name* type) {
         CHECK_NULL(typeSpec);
         map.emplace(type, typeSpec);
         return false;
+    } else if (decl->is<IR::Type_Newtype>()) {
```

Comment:
I'll look at this PR more in details when I come back from vacation on Wednesday, but I just gave it a quick look and it seems to me that you are not generating P4Info correctly here. You are expected to use the `new_type` field for the type spec (https://github.com/p4lang/p4runtime/blob/master/proto/p4/config/v1/p4types.proto#L77) and the `new_types` map need to be updated appropriately (https://github.com/p4lang/p4runtime/blob/master/proto/p4/config/v1/p4types.proto#L60).

Reply:
This is the guidance I was looking for - do we annotate PortId_t or define the new_types which you have alluded to above.  Thanks for replying during your vacation.

Reply:
Please take a look at the updated typeSpecConvertor.cpp for a step in the right direction for typeSpec.  I suspect I am missing preparing some other data.

Reply:
this looks ok, but can't we make this more uniform with other cases (like struct)?
we can define a `preorder` overload for `IR::Type_Newtype` and let that overload populate `map`

Reply:
actually, it seems that you are still not populating the `new_types` Protobuf map...

Reply:
I changed the code in typeSpecConverted.cpp and ran a test.

```hemant@ubuntu:/tmp/p4c/build$ p4/testdata/p4_16_samples/psa-example-digest-bmv2.p4.test```

The PortId_t `type` shows up as new_type.  I am not sure why in this p4info we need new_types.  If we do, could you please show where does it go?

```
  structs {
    key: "mac_learn_digest_t"
    value {
      members {
        name: "srcAddr"
        type_spec {
          bitstring {
            bit {
              bitwidth: 48
            }
          }
        }
      }
      members {
        name: "ingress_port"
        type_spec {
          new_type {
            name: "PortId_t"
          }
        }
      }
    }
  }
}
```

Reply:
`new_types` maps custom type names to a `P4NewTypeSpec` message. This message gives us necessary information about the type - in particular its bitwidth (native or translated). Please refer to the P4Runtime spec, I'm pretty sure there is a detailed description of `P4NewTypeSpec` there.

Reply:
@hesingh Here are some notes on new_type and how it affects p4info file contents.  It is not necessarily 100% complete or correct, but should be a good start to learning about it: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md

Comments, questions, corrections on that document all welcome.

Reply:
@antoninbas Please take a look at the new code checked in for preorder Type_Newtype.  The p4info still does not contain new_types.  This file has 3-4 different ways for how to populate the map for structs, enums, etc.  This is why I am not sure which way to use to generate the map for new_types.

Reply:
I figured out generating new_types.  Never mind. I changed the code and the new_types block is generated.

Reply:
I have pinged new_types and checked in the code.  I will look into adding the other translation type and making the code robust next.  

```
  new_types {
    key: "PortId_t"
    value {
      original_type {
        bitstring {
          bit {
            bitwidth: 32
          }
        }
      }
    }
  }
```

Reply:
The typeSpecConverter.cpp code is ready for review.  It has passed testing for original/translated type used with PortId_t.